### PR TITLE
Improve click & collect and stock allocation

### PR DESCRIPTION
**Pull request by IKarbowiak**
- Add many to many relations between `Channel` and `Warehouse`
- Allow assigning to the `shipping zone` only `warehouses` that have at least one common channel with the `shipping zone`
- Removed no longer valid `shipping zone - warehouse` relations (look up for `shipping zone - warehouse` that does not have a common channel anymore) when:
  - `channel` is deleted
  -  the `warehouses` from the channel are removed
  -  the `shipping zones` from the channel are removed
  - the `channels` from the shipping zone are deleted
- Take into account only `channel` when resolving available click and collect warehouses - the shipping address is not taken into consideration
- Update stock allocation - firstly allocate stocks in stocks with the highest quantity available
- Update method for getting available stocks - stock is available for a given variant when:
  - the related warehouse is available in the order/checkout channel 
  - related warehouse has a shipping zone that includes the shipping address country code and and this shipping zone is available in a given channel
- Allow filtering `warehouses` by `channels`

Should be merged together with https://github.com/saleor/saleor-plugin-avatax-excise/pull/28.
Docs: https://github.com/saleor/saleor-docs/pull/450
Spec: https://github.com/saleor/saleor/pull/9970

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-7213 Improved click&collect and stock allocation](https://app.clickup.com/t/2549495/SALEOR-7213)

File: saleor/graphql/schema.graphql [hunk 1]
```diff
@@ -3593,6 +3593,9 @@ type Channel implements Node {
   Added in Saleor 3.1.
   """
   defaultCountry: CountryDisplay!
+
+  """List of warehouses assigned to this channel."""
```
Comment (korycins) [inline saleor/graphql/schema.graphql:3597]: Shouldn't we add a label `ADDED_IN_35` for  all new fields? 👇

**Review by korycins — APPROVED** [review]

File: saleor/graphql/channel/dataloaders.py [hunk 2]
```diff
@@ -113,3 +115,33 @@ def map_shipping_zones(shipping_zones):
             .load_many({pk for pk, _ in zone_and_channel_is_pairs})
             .then(map_shipping_zones)
         )
+
+
+class WarehousesByChannelIdLoader(DataLoader):
```
Comment (fowczarek) [inline saleor/graphql/channel/dataloaders.py:120]: Should we put this dataloader to `saleor/graphql/warehouse/dataloaders.py`

File: saleor/graphql/channel/dataloaders.py [hunk 3]
```diff
@@ -113,3 +115,33 @@ def map_shipping_zones(shipping_zones):
...
+    context_key = "warehouse_by_channel"
+
+    def batch_load(self, keys):
+        warehouse_and_channel_in_pairs = (
+            Warehouse.objects.using(self.database_connection_name)
+            .filter(channels__id__in=keys)
+            .values_list("pk", "channels__id")
+        )
```
Comment (fowczarek) [inline saleor/graphql/channel/dataloaders.py:128]: This soultion should be better for database. 
```suggestion
        WarehouseChannel = Warehouse.channels.through
        warehouse_and_channel_in_pairs = (
            WarehouseChannel.objects.using(self.database_connection_name)
            .filter(channel_id__in=keys)
            .values_list("warehouse_id", "channel_id")
        )
```

**Review by fowczarek — APPROVED** [review]

File: saleor/graphql/channel/dataloaders.py (hunk 3 above)
Comment (IKarbowiak) [reply saleor/graphql/channel/dataloaders.py:128]: Good point. Thank you!

**Comment by maarcingebala** [comment]
Blocked - we're waiting for the dashboard part to be finished.

**Review by korycins — APPROVED** [review]

**Review by jakubkuc — APPROVED** [review]

**Pull request merged**